ocaml/testsuite/tests/typing-layouts-arrays/basics.ml

```diff
@@ -0,0 +1,202 @@
+(* TEST
+   * flambda2
+   ** expect
+     flags = "-extension layouts_alpha"
+   ** expect
+     flags = "-extension layouts_beta"
+*)
+
+(*******************************************)
+(* Test 1: Support unboxed types in arrays *)
+
+type t_any : any
+
+type t1 = float# array
+type t2 = int32# array
+type t3 = int64# array
+type t4 = nativeint# array
+type t5 = t_any array
+
+type ('a : float64) t1' = 'a array
+type ('a : bits32) t2' = 'a array
+type ('a : bits64) t3' = 'a array
+type ('a : word) t4' = 'a array
+type ('a : any) t5' = 'a array
+
+[%%expect{|
+type t_any : any
+type t1 = float# array
+type t2 = int32# array
+type t3 = int64# array
+type t4 = nativeint# array
+type t5 = t_any array
+type ('a : float64) t1' = 'a array
+type ('a : bits32) t2' = 'a array
+type ('a : bits64) t3' = 'a array
+type ('a : word) t4' = 'a array
+type ('a : any) t5' = 'a array
+|}];;
+
+(*****************************)
+(* Test 2: array expressions *)
+
+let v1 = [| #1. |]
+[%%expect{|
+val v1 : float# array = [|<abstr>|]
+|}];;
+
+
+let v2 = [| #1l |]
+[%%expect{|
+val v2 : int32# array = [|<abstr>|]
+|}];;
+
+
+let v3 = [| #1L |]
+[%%expect{|
+val v3 : int64# array = [|<abstr>|]
+|}];;
+
+
+let v4 = [| #1n |]
+[%%expect{|
+val v4 : nativeint# array = [|<abstr>|]
+|}];;
+
+(****************************************)
+(* Test 3: Array operations do not work *)
+
+let f (x : float# array) = x.(0)
+[%%expect{|
+Line 1, characters 27-28:
+1 | let f (x : float# array) = x.(0)
+                               ^
+Error: This expression has type float# array
+       but an expression was expected of type 'a array
+       The layout of float# is float64, because
+         it is the primitive float64 type float#.
+       But the layout of float# must be a sublayout of value, because
+         of layout requirements from an imported definition.
+|}];;
+
+let f (x : float# array) = Array.length x
+[%%expect{|
+Line 1, characters 40-41:
+1 | let f (x : float# array) = Array.length x
+                                            ^
+Error: This expression has type float# array
+       but an expression was expected of type 'a array
+       The layout of float# is float64, because
+         it is the primitive float64 type float#.
+       But the layout of float# must be a sublayout of value, because
+         of layout requirements from an imported definition.
+|}];;
+
+(*****************************************************************)
+(* Test 4: Calling wrong primitives on unboxed array kinds fails *)
+
+external get : float# array -> int -> float = "%floatarray_safe_get"
+let d (x : float# array) = get x 0
+
+[%%expect{|
+external get : float# array -> int -> float = "%floatarray_safe_get"
+Line 2, characters 27-34:
+2 | let d (x : float# array) = get x 0
+                               ^^^^^^^
+Error: Array kind unboxed_float can only be operated on using its own primitives
+       and those primitives can only work on unboxed_float
```

Comment:
I'm surprised to see such an error at all. Is it just because `[@layout_poly]` didn't exist when this PR was originally written? Letting users use the ordinary array primitives is one of the main reasons to have `float# array`. Without that it's not really better than `float_array`.

Reply:
Yes, this error message is written before the `[@layout_poly]` work and should be reworded.

The currently plan is to not introduce any new array primitives for arrays of unboxed types. Instead, we use the same ordinary array primitives with `[@layout_poly]`.

For this specific test case, it should still be an error because it's using `%floatarray_safe_get` instead of `%array_safe_get`:

```
external get : float# array -> int -> float = "%floatarray_safe_get"
let d (x : float# array) = get x 0
```

The error message should probably be something along the lines of `array containing ... can't be operated on by primitive ...`.

Reply:
Reworded the error to what's suggested: "Arrays of unboxed types can not be operated on by primitives for arrays of boxed types."

Reply:
Thanks.  But now that I understand there is one version of the primitive that works for both boxed and unboxed types, I think this error (that I suggested!) is wrong.

I believe the following:
- We have two classes of array primitives: the generic ones, which are parameterized by a kind, and the float ones.  Since float arrays and unboxed float arrays have the same runtime representation, one might wonder why we need the float ones.  The answer is that, e.g., float array set must take a boxed float as an argument and unbox it, while float# array set takes an already-unboxed float.
- The `glb_array_...` functions can, in principal, issue an error in two cases: (a) we are using a generic array primitive on a float array (and vice versa), and (b) we are using the generic array primitives on an array that has (non-float) values or unboxed types, but there is a mismatch in the kind that's recorded in the primitive and the one derived from the type of the array.
- But actually, only case (a) is reachable.  Because currently the only way to get the kind stored in the generic array primitives to be anything other than value is as the result of `glb_array_...` in the special cases where we allow value on one side an an unboxed type on the other.  And if you somehow reused a primitive after it has been specialized, you would get a type error before you reached this error in lambda.

So: I think we should revise this error to be spefically about using (boxed) float primitives on non (boxed) float arrays (wording tbd) and do a `Misc.fatal_error` or something in the cases we think are unreachable.

Reply:
Agree with the conclusion above. Changed the error message to:

```
Error: Floatarray primitives can't be used on arrays containing
       unboxed types.
```

And for case that we believe to be unreachable, `Misc.fatal_error` is used.

Reply:
Thanks.  A couple additional curiosities I am just documenting here for posterity:

* One might wonder whether there should also be an error in the opposite direction (using unboxed primitives on float arrays).  But, of course, there aren't really unboxed primitives - just normal primitives that work on unboxed kinds.  Any attempt to get such a primitive and use it on a float array would be a type error before reaching this function.

* On the other hand, this code is actually accepted and works fine:
```ocaml
external get : float array -> int -> float# = "%array_safe_get"
let d (x : float array) = get x 0
```
That is, the compiler is figuring out it doesn't need to add a box to the result of `get`, but would add a box if we had specified `float` as its return type.  This behavior pre-dates this PR, and the middle-end devs are checking that it is implmented sensibly.

Reply:
It's actually a fun story: the function thinks it needs to return the boxed result in the associated integer register, but still leaves the unboxed version in the float register. The caller thinks it returned an unboxed value in the float register, and ignores the integer register. This makes it almost work, although of course slightly more aggressive inlining settings the middle-end recognizes that something is wrong and produces a fatal error (at compilation time).